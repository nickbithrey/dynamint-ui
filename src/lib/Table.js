import React from 'react';
import PropTypes from 'prop-types';
import { DetailsList, DetailsListLayoutMode, SelectionMode } from 'office-ui-fabric-react/lib/DetailsList';

export default class Table extends React.Component {
	
	constructor(props) {
		super(props);
		this.load = props.load;
	}

	loadTable() {
		if (!this.props.list) {
			this.load();
		}
	}
	
	componentDidMount() {
		this.loadTable();
	}
	
	componentDidUpdate() {
		this.loadTable();
	}
	
	render() {
		let {
			list,
			columns
		} = this.props;
		
		if (!list) {
			return <h1>no data</h1>
		}
		return (
			<DetailsList
	            items={list}
	            compact={false}
	            columns={columns}
	            selectionMode={SelectionMode.single}
	            setKey="set"
	            layoutMode={DetailsListLayoutMode.justified}
	            isHeaderVisible={true}
				selection={this.props.selection}
	            selectionPreservedOnEmptyClick={true}
	            enterModalSelectionOnTouch={true}
	            ariaLabelForSelectionColumn="Toggle selection"
	            ariaLabelForSelectAllCheckbox="Toggle selection for all items"
	          />
		);
	}
	
}

Table.propTypes = {
	load: PropTypes.func.isRequired,
	list: PropTypes.arrayOf(PropTypes.object)
}


