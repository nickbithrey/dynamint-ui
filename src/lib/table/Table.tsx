import * as React from 'react';
import { 
    DetailsListLayoutMode, 
    IColumn, 
    DetailsRow,
    IDetailsRowProps,
    IDetailsRowBaseProps,
    DetailsColumnBase
} from 'office-ui-fabric-react/lib/DetailsList';
import { IRenderFunction } from 'office-ui-fabric-react/lib/Utilities';
import { ShimmeredDetailsList } from 'office-ui-fabric-react/lib/ShimmeredDetailsList';
import {
    DefaultTableRow,
    IDefaultTableRow
} from './DefaultTableRow';
import { ISelection } from 'office-ui-fabric-react/lib/Selection';

export interface ITable {
    loading: boolean;
    items: Array<ITableItem>;
    columns: Array<ITableColumn>;
    selection?: ISelection;
    onRenderRow?: IRenderFunction<IDefaultTableRow>;
    onRenderItemColumn?: (item: ITableItem, index: number, column: ITableColumn) => React.ReactNode;
    onItemInvoked?: (item?: ITableItem, index?: number, ev?: Event) => void;
}

export interface ITableItem {
    [key: string]: any;
}

export interface ITableColumn extends IColumn {
    name: string;
    key: string;
    fieldName: string;
    minWidth: number;
    maxWidth?: number;
}

export class Table extends React.Component<ITable> {
    
    constructor(props: ITable) {
        super(props);
    }
    
    render() {
        const {
            items,
            columns,
            loading,
            selection,
            onRenderRow,
            onRenderItemColumn,
            onItemInvoked
        } = this.props;
        return (
                <ShimmeredDetailsList
                    items={items ? items : []}
                    columns={columns}
                    enableShimmer={loading}
                    selection={selection}
                    selectionPreservedOnEmptyClick={true}
                    onRenderRow={onRenderRow}
                    onRenderItemColumn={onRenderItemColumn}
                    onItemInvoked={onItemInvoked}
                />
        );
    }
    
}