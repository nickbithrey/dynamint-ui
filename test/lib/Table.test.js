import React from 'react';
import { shallow } from 'enzyme';

import Table from '~/lib/Table';

const Record = props => (
	<tr>
		<td>{props.reference}</td>
		<td>{props.description}</td>
	</tr>
);

describe('<Table /> rendering', () => {

	it('can render standard table with records', () => {
		const list = [
			{reference:'ref', description: 'desc'},
			{reference:'ref2', description: 'desc2'}
		]
		const load = jest.fn(() => []);
		const wrapper = shallow( <Table list={list} load={load} /> );
		
		expect(wrapper).toMatchSnapshot();
		expect(wrapper.instance().props.load).not.toBeCalled();
	});
	
	it('can render empty table when no values passed into the list', () => {
		const load = jest.fn(() => []);
		const wrapper = shallow( <Table list={[]} load={load} tag={Record} /> );

		expect(wrapper.find('StyledWithViewportComponent')).toHaveLength(0);
		
		const noData = wrapper.find('h1');
		expect(noData.text()).toEqual('no data');
	});
		
	it('renders table with same number of records and in list', () => {
		const list = [
			{},
			{},
			{}
		]
		const load = jest.fn(() => []);
		const wrapper = shallow( <Table list={list} load={load} tag={Record} /> );
		
		expect(wrapper.find('StyledWithViewportComponent')).toHaveLength(1);
		const listProps = wrapper.find('StyledWithViewportComponent').props();
		expect(listProps.items).toHaveLength(list.length);
		expect(wrapper.instance().props.load).not.toBeCalled();
	});
		
	it('calls load function when list is undefined on mount', () => {
		const load = jest.fn(() => []);
		const wrapper = shallow( <Table load={load} tag={Record} /> );
		
		expect(wrapper.instance().props.load).toBeCalled();
	});
		
	it('calls load function when list is undefined on update', () => {
		const load = jest.fn(() => []);
		const wrapper = shallow( <Table list={[]} load={load} tag={Record} /> );

		// assert load not called already on mount (as list is defined)
		expect(wrapper.instance().props.load).not.toBeCalled();
		
		// update component
		wrapper.setProps({list: null});
		// assert load called on update
		expect(wrapper.instance().props.load).toBeCalled();
	});
		
});