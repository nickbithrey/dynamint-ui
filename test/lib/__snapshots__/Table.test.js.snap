// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Table /> rendering can render standard table with records 1`] = `
<StyledWithViewportComponent
  ariaLabelForSelectAllCheckbox="Toggle selection for all items"
  ariaLabelForSelectionColumn="Toggle selection"
  compact={false}
  enterModalSelectionOnTouch={true}
  isHeaderVisible={true}
  items={
    Array [
      Object {
        "description": "desc",
        "reference": "ref",
      },
      Object {
        "description": "desc2",
        "reference": "ref2",
      },
    ]
  }
  layoutMode={1}
  selectionMode={1}
  selectionPreservedOnEmptyClick={true}
  setKey="set"
/>
`;
